### PR TITLE
Adds Batch client connection and async RPC calls to the bitcoind client

### PR DESCRIPTION
**Pull request by Vib-UX**
# PR serves in relation to fixing issue [5041](https://github.com/lightningnetwork/lnd/issues/5041#issue-810611891)

With this PR we enable batch client connection and async RPC calls to the bitcoind client providing testcoverages for the same. This helps us in batching the required requests needed for speeding up operations by reducing round trips. 

### Testing
* [ ]  Your PR passes all CI checks.
* [x]  Tests covering the positive and negative (error paths) are included.
* [ ]  Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
* [x]  The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
* [x]  Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).

📝 Please see our [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

### CODE REVIEW

File: chain/bitcoind_client.go [hunk 1]
```diff
@@ -150,6 +151,18 @@ func (c *BitcoindClient) GetBlockVerbose(
 	return c.chainConn.client.GetBlockVerbose(hash)
 }
 
+// GetBestBlockHashAsync returns an instance of a type that can be used to get
+// the result of the RPC at some future time by invoking the Receive function on
+// the returned instance.
+func (c *BitcoindClient) GetBlockHashAsync(height int64) (rpcclient.FutureGetBlockHashResult, error) {
```
Comment (guggero) [inline chain/bitcoind_client.go]: nit: line too long, should break before `height` parameter and then add a newline before the method body (since this then counts as a multi-line method declaration). See https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#code-spacing-and-formatting for general styling information, we try to adhere to those in the `btcsuite` projects as well (even though a lot of the existing code doesn't conform yet).

File: chain/bitcoind_client.go [hunk 2]
```diff
@@ -150,6 +151,18 @@ func (c *BitcoindClient) GetBlockVerbose(
...
+	return c.chainConn.batchClient.GetBlockHashAsync(height), nil
+}
+
+// func to send bulk request
```
Comment (guggero) [inline chain/bitcoind_client.go]: nit: godoc comment must start with method name and should be full sentence.

File: chain/bitcoind_conn.go [hunk 3]
```diff
@@ -101,6 +101,9 @@ type BitcoindConn struct {
 	// client is the RPC client to the bitcoind node.
 	client *rpcclient.Client
 
+	// batch-Client is the RPC client to the bitcoind node
```
Comment (guggero) [inline chain/bitcoind_conn.go]: nit: a godoc comment should start with the exact variable name, so this should say `batchClient` instead of `batch-Client`.

File: chain/bitcoind_client.go [hunk 4]
```diff
@@ -150,6 +151,18 @@ func (c *BitcoindClient) GetBlockVerbose(
...
+func (c *BitcoindClient) Send() error {
```
Comment (guggero) [inline chain/bitcoind_client.go]: Should this method be called `SendAsyncQueue` perhaps? Otherwise in this context just "Send" doesn't really say what it's doing.

**Review by guggero — COMMENTED** [review]
Looks pretty good! Mostly style nits, functionality looks to there :+1:

File: chain/bitcoind_client.go (hunk 4 above)
Comment (bhandras) [reply chain/bitcoind_client.go]: +1, it's worth thinking in less simplicity and more towards readability of the code, where more readable names tend to be not too long and not too short :)

File: chain/bitcoind_conn.go (hunk 3 above)
Comment (bhandras) [reply chain/bitcoind_conn.go]: nit: comments should always end with a punctuation.

File: chain/bitcoind_conn.go [hunk 5]
```diff
@@ -101,7 +101,7 @@ type BitcoindConn struct {
 	// client is the RPC client to the bitcoind node.
 	client *rpcclient.Client
 
-	// batch-Client is the RPC client to the bitcoind node
+	// batchClient is the RPC client to the bitcoind node.
```
Comment (bhandras) [inline chain/bitcoind_conn.go]: Please squash fixup commits, so that the pull request always consists of close to final state.

File: chain/bitcoind_client.go [hunk 6]
```diff
@@ -381,16 +392,33 @@ func (c *BitcoindClient) RescanBlocks(
 	blockHashes []chainhash.Hash) ([]btcjson.RescannedBlock, error) {
 
 	rescannedBlocks := make([]btcjson.RescannedBlock, 0, len(blockHashes))
+
+	// request map will store rpcclient.FutureGetBlockHeaderVerboseResult for
+	// each blockHash.
+	request := make(
+		map[chainhash.Hash]rpcclient.FutureGetBlockHeaderVerboseResult, len(
```
Comment (bhandras) [inline chain/bitcoind_client.go]: nit: formatting

File: chain/bitcoind_client.go [hunk 7]
```diff
@@ -381,16 +392,33 @@ func (c *BitcoindClient) RescanBlocks(
...
+			blockHashes))
+
 	for _, hash := range blockHashes {
 		hash := hash
 
-		header, err := c.GetBlockHeaderVerbose(&hash)
+		headerRequest, err := c.GetBlockHeaderVerboseAsync(&hash)
+		if err != nil {
+			log.Warnf("Unable to receive header promise: %s", err)
```
Comment (bhandras) [inline chain/bitcoind_client.go]: nit: could log the hash too.

File: chain/bitcoind_client.go [hunk 8]
```diff
@@ -381,16 +392,33 @@ func (c *BitcoindClient) RescanBlocks(
...
+			continue
+		}
+		request[hash] = headerRequest
+	}
+
+	c.SendAsyncQueue()
+
+	for hash, headerRequest := range request {
```
Comment (bhandras) [inline chain/bitcoind_client.go]: nit: since `hash` is a local variable tracking the current element, it's worth adding `hash := hash` below this line such that any functions where a pointer to hash is passed don't end up tracking the last element in `requests`.

File: chain/bitcoind_client.go [hunk 9]
```diff
@@ -419,6 +419,7 @@ func (c *BitcoindClient) RescanBlocks(
 				hash, err)
 			continue
 		}
+
```
Comment (bhandras) [inline chain/bitcoind_client.go:649]: Squash with previous commit.

File: chain/bitcoind_client.go [hunk 10]
```diff
@@ -962,17 +962,21 @@ func (c *BitcoindClient) FilterBlocks(
 
 	blockFilterer := NewBlockFilterer(c.chainConn.cfg.ChainParams, req)
 
+	storeHash := make([]*chainhash.Hash, len(req.Blocks))
```
Comment (bhandras) [inline chain/bitcoind_client.go]: The correct way to declare this slice would be `make([]*chainHash.Hash, 0, len(req.Blocks))` otherwise it'll have a predefined size and then we'll also append `len(req.Blocks)` elements to it.

File: chain/bitcoind_client.go [hunk 11]
```diff
@@ -962,17 +962,21 @@ func (c *BitcoindClient) FilterBlocks(
...
+
+	for _, block := range req.Blocks {
```
Comment (bhandras) [inline chain/bitcoind_client.go]: Same scope bug as above. Here's a quick example of what's happening: https://go.dev/play/p/Q7OdkHdPa_g
If you run that example you'll see that we're storing the pointer to the same element every time.

File: chain/bitcoind_conn.go [hunk 12]
```diff
@@ -600,6 +600,70 @@ func (c *BitcoindConn) GetBlock(hash *chainhash.Hash) (*wire.MsgBlock, error) {
 	}
 }
 
+// GetBlocksBatch returns a batch of raw blocks from the server given their
+// hashes. If the server has already pruned some of the blocks, it will be
+// retrieved from their peers.
+func (c *BitcoindConn) GetBlocksBatch(
+	hashes []*chainhash.Hash) ([]*wire.MsgBlock, error) {
+
+	batchRequest := make(map[*chainhash.Hash]rpcclient.FutureGetBlockResult,
```
Comment (bhandras) [inline chain/bitcoind_conn.go]: nit: formatting.

File: chain/bitcoind_conn.go [hunk 13]
```diff
@@ -600,6 +600,70 @@ func (c *BitcoindConn) GetBlock(hash *chainhash.Hash) (*wire.MsgBlock, error) {
...
+		len(hashes))
+
+	batchRawBlocks := make([]*wire.MsgBlock, 0, len(hashes))
```
Comment (bhandras) [inline chain/bitcoind_conn.go]: nit: move declaration closer to usage.

File: chain/bitcoind_conn.go [hunk 14]
```diff
@@ -600,6 +600,70 @@ func (c *BitcoindConn) GetBlock(hash *chainhash.Hash) (*wire.MsgBlock, error) {
...
+				// for the block now.
+
+			case <-c.quit:
+				return nil, ErrBitcoindClientShuttingDown
+			}
+		}
+	}
+
+	return batchRawBlocks, nil
+
```
Comment (bhandras) [inline chain/bitcoind_conn.go]: nit: extra nl

File: chain/bitcoind_conn.go [hunk 15]
```diff
@@ -600,6 +600,70 @@ func (c *BitcoindConn) GetBlock(hash *chainhash.Hash) (*wire.MsgBlock, error) {
...
+	batchRequest := make(map[*chainhash.Hash]rpcclient.FutureGetBlockResult,
+		len(hashes))
+
+	batchRawBlocks := make([]*wire.MsgBlock, 0, len(hashes))
+	for _, hash := range hashes {
+		sendRequest := c.batchClient.GetBlockAsync(hash)
+		batchRequest[hash] = sendRequest
+	}
+
+	c.batchClient.Send()
```
Comment (bhandras) [inline chain/bitcoind_conn.go]: Returned error not checked.

**Review by bhandras — COMMENTED** [review]
Nice work @Vib-UX :zap: One general comment I can add is to please add some more detail to your commit messages. In general a good commit message is similar to what's describe in this doc:https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#model-git-commit-messages

Obviously it's not a requirement, but a good practice. At a minimum add some short description about what your commit changes, eg. `packagename: adding a new shiny func to achieve some goals.`

Looking forward for some test coverage as well :)

File: chain/bitcoind_conn.go (hunk 5 above)
Comment (Vib-UX) [reply chain/bitcoind_conn.go]: Sure

File: chain/bitcoind_client.go (hunk 7 above)
Comment (Vib-UX) [reply chain/bitcoind_client.go]: Yes

File: chain/bitcoind_client.go (hunk 8 above)
Comment (Vib-UX) [reply chain/bitcoind_client.go]: Thanks for the explanation, thats insightful!

File: chain/bitcoind_client.go (hunk 11 above)
Comment (Vib-UX) [reply chain/bitcoind_client.go]: Attached example was insightful,
https://medium.com/swlh/use-pointer-of-for-range-loop-variable-in-go-3d3481f7ffc9

Iterating variable with `range` re-assigns value to the same address location. Got all my doubts clear by reading the above blog, thanks for pointing it out.

File: chain/bitcoind_client.go [hunk 16]
```diff
@@ -366,10 +393,31 @@ func (c *BitcoindClient) RescanBlocks(
 	blockHashes []chainhash.Hash) ([]btcjson.RescannedBlock, error) {
 
 	rescannedBlocks := make([]btcjson.RescannedBlock, 0, len(blockHashes))
+
+	// request map will store rpcclient.FutureGetBlockHeaderVerboseResult for
+	// each blockHash.
+	request := make(
```
Comment (bhandras) [inline chain/bitcoind_client.go]: nit: I think renaming this to `requests` is slightly better. Also fix formatting.

File: chain/bitcoind_client.go [hunk 17]
```diff
@@ -366,10 +393,31 @@ func (c *BitcoindClient) RescanBlocks(
...
+		map[chainhash.Hash]rpcclient.FutureGetBlockHeaderVerboseResult,
+		len(blockHashes))
+
 	for _, hash := range blockHashes {
 		hash := hash
 
-		header, err := c.GetBlockHeaderVerbose(&hash)
+		headerRequest, err := c.GetBlockHeaderVerboseAsync(&hash)
+		if err != nil {
+			log.Warnf(
```
Comment (bhandras) [inline chain/bitcoind_client.go]: nit: formatting

File: chain/bitcoind_client.go [hunk 18]
```diff
@@ -366,10 +393,31 @@ func (c *BitcoindClient) RescanBlocks(
...
+		}
+		request[hash] = headerRequest
+	}
+
+	c.SendAsyncQueue()
+
+	for hash, headerRequest := range request {
+		hash := hash
+
+		header, err := headerRequest.Receive()
```
Comment (bhandras) [inline chain/bitcoind_client.go]: Could this function also use `GetBlocksBatch` below?

File: chain/bitcoind_conn.go [hunk 19]
```diff
@@ -425,6 +436,73 @@ func (c *BitcoindConn) GetBlock(hash *chainhash.Hash) (*wire.MsgBlock, error) {
 	}
 }
 
+// GetBlocksBatch returns a batch of raw blocks from the server given their
+// hashes. If the server has already pruned some of the blocks, it will be
+// retrieved from their peers.
+func (c *BitcoindConn) GetBlocksBatch(
+	hashes []*chainhash.Hash) ([]*wire.MsgBlock, error) {
+
+	batchRequest := make(
```
Comment (bhandras) [inline chain/bitcoind_conn.go]: nit: formatting

File: chain/bitcoind_client.go [hunk 20]
```diff
@@ -12,6 +12,7 @@ import (
 	"github.com/btcsuite/btcd/btcjson"
 	"github.com/btcsuite/btcd/btcutil"
 	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/rpcclient"
```
Comment (bhandras) [inline chain/bitcoind_client.go:15]: nit: commit messages should be formatted such that the first line includes a short message then a line is kept empty, and finally you add a longer explanation. In this case something like:

```
chain: enable batched operations with bitcoind

In this commit we add support for batched operations
when using the bitcoind client.
```

File: chain/bitcoind_conn.go [hunk 21]
```diff
@@ -443,7 +443,9 @@ func (c *BitcoindConn) GetBlocksBatch(
 	hashes []*chainhash.Hash) ([]*wire.MsgBlock, error) {
 
 	batchRequest := make(
-		map[*chainhash.Hash]rpcclient.FutureGetBlockResult, len(hashes))
+		map[*chainhash.Hash]rpcclient.FutureGetBlockResult,
```
Comment (bhandras) [inline chain/bitcoind_conn.go]: Changes to `GetBlocksBatch` should be in the previous commit.

File: chain/bitcoind_conn.go [hunk 22]
```diff
@@ -469,8 +471,7 @@ func (c *BitcoindConn) GetBlocksBatch(
 		// it wasn't due to the block being pruned, add nil entry for batchIndex
 		// and continue to the next request.
 		if !isBlockPrunedErr(err) || c.prunedBlockDispatcher == nil {
-			batchRawBlocks = append(batchRawBlocks, nil)
-			continue
+			return nil, err
```
Comment (bhandras) [inline chain/bitcoind_conn.go:475]: Make sure to also update the above comment (in the previous commit once you squash these changes).

File: chain/bitcoind_client.go (hunk 20 above)
Comment (bhandras) [inline chain/bitcoind_client.go:15]: nit: I'd rephrase your commit message to:

```
chain: add batched async calls support for bitcoind client

This commit adds a batch client connection and async RPC
calls to the bitcoind client which will enable batching multiple
requests speeding up operations by reducing round trips.
```

File: chain/bitcoind_conn.go [hunk 23]
```diff
@@ -94,6 +94,9 @@ type BitcoindConn struct {
 	// client is the RPC client to the bitcoind node.
 	client *rpcclient.Client
 
+	// batchClient is the RPC client to the bitcoind node.
```
Comment (bhandras) [inline chain/bitcoind_conn.go]: nit: 
```
// batchClient is the batched RPC client to the bitcoind node.
```

File: chain/bitcoind_conn.go [hunk 24]
```diff
@@ -234,6 +243,8 @@ func (c *BitcoindConn) Stop() {
 		log.Errorf("error shutting down bitcoind events: %w", err)
 	}
 
+	c.batchClient.Shutdown()
```
Comment (bhandras) [inline chain/bitcoind_conn.go:250]: Do we also need `c.batchClient.WaitForShutdown()` like we do for the normal client (line 252).

File: chain/bitcoind_client.go [hunk 25]
```diff
@@ -160,6 +160,13 @@ func (c *BitcoindClient) GetBlock(hash *chainhash.Hash) (*wire.MsgBlock, error)
 	return c.chainConn.GetBlock(hash)
 }
 
+// GetBlocksBatch returns batchBlocks from batchHashes.
```
Comment (bhandras) [inline chain/bitcoind_client.go]: nit: commit message:
```
chain: add GetBlocksBatch() to the bitcoind client

This will leverage the updated batch client to fetch multiple blocks in
one go which can speed up rescan and filter blocks.
```

File: chain/bitcoind_client.go [hunk 26]
```diff
@@ -427,10 +427,44 @@ func (c *BitcoindClient) RescanBlocks(
 	blockHashes []chainhash.Hash) ([]btcjson.RescannedBlock, error) {
 
 	rescannedBlocks := make([]btcjson.RescannedBlock, 0, len(blockHashes))
+
```
Comment (bhandras) [inline chain/bitcoind_client.go]: nit: in commit message:
```
chain: updates RescanBlocks() & FilterBlocks()

Both RescanBlocks() and FilterBlocks() now leverages GetBlocksBatch()
which works with the updated batch client for speed ups.
```

File: chain/bitcoind_client.go [hunk 27]
```diff
@@ -427,10 +427,44 @@ func (c *BitcoindClient) RescanBlocks(
...
+		len(blockHashes),
+	)
+
+	for _, hash := range blockHashes {
+		hash := hash
+
+		headerRequest, err := c.GetBlockHeaderVerboseAsync(&hash)
+		if err != nil {
+			log.Warnf(
+				"Unable to receive header promise from hash %v: %s", hash, err)
```
Comment (bhandras) [inline chain/bitcoind_client.go]: nit: format to 80 columns.

File: chain/bitcoind_client.go [hunk 28]
```diff
@@ -427,10 +427,44 @@ func (c *BitcoindClient) RescanBlocks(
...
+			continue
+		}
+		requests[hash] = headerRequest
+	}
+
+	c.SendAsyncQueue()
```
Comment (bhandras) [inline chain/bitcoind_client.go]: I think this function could be reworked a bit from its current form, to:

1) fetch block headers to get timestamps and heights (send async and receive too)
2) in the receive loop you could filter out those which are not important for us:
```
if !c.shouldFilterBlock(time.Unix(header.Time, 0)) {
			continue
}
```
3) finally get the filtered block hashes

File: chain/bitcoind_client.go [hunk 29]
```diff
@@ -981,17 +1008,21 @@ func (c *BitcoindClient) FilterBlocks(
 
 	blockFilterer := NewBlockFilterer(c.chainConn.cfg.ChainParams, req)
 
+	storeHash := make([]*chainhash.Hash, 0, len(req.Blocks))
```
Comment (bhandras) [inline chain/bitcoind_client.go]: nit: rename `storeHash` to `blockHashes`

File: chain/bitcoind_client.go [hunk 30]
```diff
@@ -981,17 +1008,21 @@ func (c *BitcoindClient) FilterBlocks(
...
+
+	for _, block := range req.Blocks {
```
Comment (bhandras) [inline chain/bitcoind_client.go]: nit: better to use the form: `for i := range req.Blocks` and then refer to the i'th block as `req.Blocks[i]`.

File: chain/bitcoind_client.go [hunk 31]
```diff
@@ -981,17 +1008,21 @@ func (c *BitcoindClient) FilterBlocks(
...
 	// above, breaking out early if any addresses are found.
-	for i, block := range req.Blocks {
-		// TODO(conner): add prefetching, since we already know we'll be
-		// fetching *every* block
-		rawBlock, err := c.GetBlock(&block.Hash)
-		if err != nil {
-			return nil, err
-		}
-
+	for i, rawBlock := range rawBlocks {
```
Comment (bhandras) [inline chain/bitcoind_client.go]: nit: same here, would not surface `rawBlock` to avoid copy.

File: chain/bitcoind_client.go [hunk 32]
```diff
@@ -1003,14 +1034,15 @@ func (c *BitcoindClient) FilterBlocks(
...
-			BlockMeta:          block,
+			BlockMeta:          req.Blocks[i],
 			FoundExternalAddrs: blockFilterer.FoundExternal,
 			FoundInternalAddrs: blockFilterer.FoundInternal,
 			FoundOutPoints:     blockFilterer.FoundOutPoints,
 			RelevantTxns:       blockFilterer.RelevantTxns,
 		}
 
 		return resp, nil
+
```
Comment (bhandras) [inline chain/bitcoind_client.go:1310]: nit: extra nl

**Review by bhandras — CHANGES_REQUESTED** [review]
Great work! Getting very close. `RescanBlocks()` needs some more fixing but other than that almost good to go.

File: chain/bitcoind_client.go [hunk 33]
```diff
@@ -426,14 +426,38 @@ func (c *BitcoindClient) LoadTxFilter(reset bool, filters ...interface{}) error
...
+	)
 
-		header, err := c.GetBlockHeaderVerbose(&hash)
+	for i := range blockHashes {
+		hash := blockHashes[i]
+		header, err := requests[hash].Receive()
 		if err != nil {
 			log.Warnf("Unable to get header %s from bitcoind: %s",
-				hash, err)
+				blockHashes[i], err)
```
Comment (bhandras) [inline chain/bitcoind_client.go]: nit: can just use `hash` here.

File: chain/bitcoind_client.go [hunk 34]
```diff
@@ -442,27 +466,53 @@ func (c *BitcoindClient) RescanBlocks(
...
+			blockHashPointers[start:end],
+		)
 		if err != nil {
-			log.Warnf("Unable to get block %s from bitcoind: %s",
-				hash, err)
-			continue
+			return nil, err
 		}
+		batchedBlockHashPointers := blockHashPointers[start:end]
+		index := -1
```
Comment (bhandras) [inline chain/bitcoind_client.go]: nit: not sure what the role of the variable `index` is?

File: chain/bitcoind_client.go [hunk 35]
```diff
@@ -442,27 +466,53 @@ func (c *BitcoindClient) RescanBlocks(
...
+			hash := batchedBlockHashPointers[j]
+			relevantTxs := c.filterBlock(
+				blocks[index], filteredHeaders[*hash].Height, false,
+			)
+			if len(relevantTxs) > 0 {
+				rescannedBlock := btcjson.RescannedBlock{
+					Hash: hash.String(),
+				}
+				for _, tx := range relevantTxs {
+					rescannedBlock.Transactions = append(
```
Comment (bhandras) [inline chain/bitcoind_client.go:691]: nit: `rescannedBlock.Transactions` can be preallocated.

File: chain/bitcoind_client.go [hunk 36]
```diff
@@ -981,36 +1031,54 @@ func (c *BitcoindClient) FilterBlocks(
...
-			FoundOutPoints:     blockFilterer.FoundOutPoints,
-			RelevantTxns:       blockFilterer.RelevantTxns,
-		}
+			// If any external or internal addresses were detected
+			// in this block, we return them to the caller so that
+			// the rescan windows can widened with subsequent addresses.
+			// The `BatchIndex` is returned so that the caller can
+			// compute the *next* block from which to begin again.
+			resp := &FilterBlocksResponse{
+				BatchIndex:         uint32(j + start),
```
Comment (bhandras) [inline chain/bitcoind_client.go]: nit: maybe `start + j` is more logical here since `start` is the offset, and `j` is the index in the batch.

File: chain/bitcoind_client_test.go [hunk 37]
```diff
@@ -0,0 +1,127 @@
...
+
+	t.Helper()
+
+	_, minerHeight, err := miner.Client.GetBestBlock()
+	require.NoError(t, err, "unable to retrieve miner's current height")
+
+	timeout := time.After(10 * time.Second)
+	for {
+		_, bitcoindHeight, err := bitcoindClient.GetBestBlock()
+		if err != nil {
```
Comment (bhandras) [inline chain/bitcoind_client_test.go]: nit: can use `require.NoError` here too.

File: chain/bitcoind_client_test.go [hunk 38]
```diff
@@ -0,0 +1,127 @@
...
+		}
+
+		if bitcoindHeight == minerHeight {
+			return uint32(bitcoindHeight)
+		}
+
+		select {
+		case <-time.After(100 * time.Millisecond):
+		case <-timeout:
+			t.Fatalf("timed out waiting to sync notifier")
```
Comment (bhandras) [inline chain/bitcoind_client_test.go]: nit: timeout message should be updated.

File: chain/bitcoind_client_test.go [hunk 39]
```diff
@@ -0,0 +1,127 @@
...
+	msgBlocksStandard := make([]*wire.MsgBlock, 0, currentHeight-broadcastHeight)
+	for i := currentHeight; i >= broadcastHeight; i-- {
+		blockHash, err := bitcoindClient.GetBlockHash(int64(i))
+		require.NoError(t, err, "unable to retrieve blockhash")
+		blockHashes = append(blockHashes, blockHash)
+		msgBlock, err := bitcoindClient.GetBlock(blockHash)
+		require.NoError(t, err)
+		msgBlocksStandard = append(msgBlocksStandard, msgBlock)
+	}
+	msgBlocksBatchClient, err := bitcoindClient.GetBlocksBatch(blockHashes)
```
Comment (bhandras) [inline chain/bitcoind_client_test.go:131]: Nice test!

File: chain/bitcoind_client_test.go [hunk 40]
```diff
@@ -125,3 +125,52 @@ func testBitcoindGetBlocksBatch(t *testing.T, rpcpolling bool) {
 		t, msgBlocksStandard, msgBlocksBatchClient, "blocks result mismatch",
 	)
 }
+
+// TestBitcoindRescanBlocks ensures that we correctly retrieve the required details
+// for the watched txn.
+func TestBitcoindRescanBlocks(t *testing.T) {
+	testBitcoindRescanBlocks(t, true)
```
Comment (bhandras) [inline chain/bitcoind_client_test.go]: nit: can create a subtest here with:

`t.Run(fmt.Sprintf("rpcpoll=true"), func(t *testing.T) { ... })` etc

File: chain/bitcoind_client_test.go [hunk 41]
```diff
@@ -125,3 +125,52 @@ func testBitcoindGetBlocksBatch(t *testing.T, rpcpolling bool) {
...
+	testBitcoindRescanBlocks(t, false)
+}
+func testBitcoindRescanBlocks(t *testing.T, rpcpolling bool) {
```
Comment (bhandras) [inline chain/bitcoind_client_test.go]: nit: add extra nl above

File: chain/bitcoind_client_test.go [hunk 42]
```diff
@@ -174,3 +176,78 @@ func testBitcoindRescanBlocks(t *testing.T, rpcpolling bool) {
...
 
 }
+
+// TestBitcoindFilterBlocks ensures that we correctly retrieve the required
+// FilterBlocksResponse for the requested watchPoints.
+func TestBitcoindFilterBlocks(t *testing.T) {
+	testBitcoindFilterBlocks(t, true)
+	testBitcoindFilterBlocks(t, false)
+}
+func testBitcoindFilterBlocks(t *testing.T, rpcpolling bool) {
```
Comment (bhandras) [inline chain/bitcoind_client_test.go]: nit: missing extra nl

**Review by bhandras — APPROVED** [review]
LGTM, just a few nits remain, great work @Vib-UX :tada:

**Comment by bhandras** [comment]
Actually one last minute comment that maybe needs more work: it is a bit unfortunate that we need to add back old code in order to test expectations, perhaps a better way would be to instead check the output of the new batched call since we know what the output should look like. Otherwise there'll always remain some unused code as part of the public interface that we'll never use.

Alternatively you can also rename the "Old" functions as "Sync", meaning they're non batched, slow sequential calls. Also what do you think @guggero ?

File: chain/bitcoind_client.go (hunk 34 above)
Comment (Vib-UX) [reply chain/bitcoind_client.go]: Yes, this is redundant now since I am traversing using index, it was there as per earlier implementation which used maps for traversal. Will fix it thanks for pointing it out!

**Comment by Vib-UX** [comment]
> Actually one last minute comment that maybe needs more work: it is a bit unfortunate that we need to add back old code in order to test expectations, perhaps a better way would be to instead check the output of the new batched call since we know what the output should look like. Otherwise there'll always remain some unused code as part of the public interface that we'll never use.
> 
> Alternatively you can also rename the "Old" functions as "Sync", meaning they're non batched, slow sequential calls. Also what do you think @guggero ?

btcwallet-testCoverageSpeedUp

The updated testcoverage removes dependency from the previous Sync functions making it redundant. It was earlier there to compare speedup.

**Comment by guggero** [comment]
> Alternatively you can also rename the "Old" functions as "Sync", meaning they're non batched, slow sequential calls. Also what do you think @guggero ?

Just jumping in to answer the question, I'm probably lacking most of the context of the previous discussion. But I'd generally not rename any existing functions, so other users (`lnd` isn't the only software that uses `btcwallet` AFAIK) don't need to change any code when updating. And using the same name as before for something that works differently is not a good idea either. So I'd keep everything existing as it is and add `*Async` methods for new stuff (which I believe is what's currently being done?).

**Comment by bhandras** [comment]
> > Alternatively you can also rename the "Old" functions as "Sync", meaning they're non batched, slow sequential calls. Also what do you think @guggero ?
> 
> Just jumping in to answer the question, I'm probably lacking most of the context of the previous discussion. But I'd generally not rename any existing functions, so other users (`lnd` isn't the only software that uses `btcwallet` AFAIK) don't need to change any code when updating. And using the same name as before for something that works differently is not a good idea either. So I'd keep everything existing as it is and add `*Async` methods for new stuff (which I believe is what's currently being done?).

Yeah maybe best to keep existing functionality and unrenamed. However since the new functionality is not exactly the same semantics as when using the "Async" versions of the RPC API, we could name them "Batched" perhaps to refer that it is doing batched calls under the hood.

**Comment by Vib-UX** [comment]
rebase master : [`a6ba11e`](https://github.com/Vib-UX/btcwallet/commit/a6ba11e2d983bc556218c65703da656367975111) to [`36bcc1e`](https://github.com/Vib-UX/btcwallet/commit/36bcc1ea2655b636937d31c7223d8c09d3569c45)

File: chain/bitcoind_conn.go [hunk 43]
```diff
@@ -437,6 +437,73 @@ func (c *BitcoindConn) GetBlock(hash *chainhash.Hash) (*wire.MsgBlock, error) {
 	}
 }
 
+// GetBlocksBatch returns a batch of raw blocks from the server given their
+// hashes. If the server has already pruned some of the blocks, it will be
+// retrieved from their peers.
+func (c *BitcoindConn) GetBlocksBatch(
+	hashes []*chainhash.Hash) ([]*wire.MsgBlock, error) {
+
+	batchRequest := make(
```
Comment (bhandras) [inline chain/bitcoind_conn.go]: nit: `batchRequest` could actually just be a slice, no need to key it with the hashes since we always iterate on `hashes`.

File: chain/bitcoind_client.go [hunk 44]
```diff
@@ -469,6 +469,103 @@ func (c *BitcoindClient) RescanBlocks(
...
 }
 
+// RescanBlocksBatched rescans any blocks passed in batched manner using asyn
+// calls, returning only the blocks that matched as []btcjson.BlockDetails.
+func (c *BitcoindClient) RescanBlocksBatched(
+	blockHashes []chainhash.Hash) ([]btcjson.RescannedBlock, error) {
+
+	// requests map will store rpcclient.FutureGetBlockHeaderVerboseResult
+	// for each blockHash.
+	requests := make(
```
Comment (bhandras) [inline chain/bitcoind_client.go:621]: nit: i think this also could just be an array since we always iterate on `blockHashes`.

File: chain/bitcoind_client.go [hunk 45]
```diff
@@ -1017,6 +1114,71 @@ func (c *BitcoindClient) FilterBlocks(
 	return nil, nil
 }
 
+// FilterBlocksBatched scans the blocks in batched manner using asyn calls
+// contained in the FilterBlocksRequest for any addresses of interest returning
+// a FilterBlocksReponse for the first block containing a matching address.
+// If no matches are found in the range of blocks requested, the returned
+// response will be nil.
+
```
Comment (bhandras) [inline chain/bitcoind_client.go]: nit: extra nl

File: chain/bitcoind_client_test.go [hunk 46]
```diff
@@ -0,0 +1,126 @@
...
+	msgBlocksStandard := make([]*wire.MsgBlock, 0, currentHeight-broadcastHeight)
+	for i := currentHeight; i >= broadcastHeight; i-- {
+		blockHash, err := bitcoindClient.GetBlockHash(int64(i))
+		require.NoError(t, err, "unable to retrieve blockhash")
+		blockHashes = append(blockHashes, blockHash)
+		msgBlock, err := bitcoindClient.GetBlock(blockHash)
+		require.NoError(t, err)
+		msgBlocksStandard = append(msgBlocksStandard, msgBlock)
+	}
+	msgBlocksBatchClient, err := bitcoindClient.GetBlocksBatch(blockHashes)
```
Comment (bhandras) [inline chain/bitcoind_client_test.go:131]: nit: add an extra nl after the for block

File: chain/bitcoind_client_test.go [hunk 47]
```diff
@@ -0,0 +1,126 @@
...
+	blockHashes := []*chainhash.Hash{&unknownHash1, &unknownHash2, &unknownHash3}
+	broadcastHeight := syncBitcoindWithMiner(t, bitcoindClient, miner)
+	_, err := bitcoindClient.GetBlocksBatch(blockHashes)
+	require.EqualError(t, err, "-5: Block not found")
+
+	// Here, we'll generate multiple valid testblocks and retrieve it
+	// linearly with GetBlock() (store the result for each GetBlock() in a
+	// slice) further also try to retrieve blocks in batch with GetBlocksBatch().
+	// At last will compare the results from both which should be equal.
+	if _, err := miner.Client.Generate(5); err != nil {
```
Comment (bhandras) [inline chain/bitcoind_client_test.go]: nit: I think it'd be better to generate something like 2x15 + some blocks to test that we do create multiple batches correctly.

**Review by bhandras — APPROVED** [review]
LGTM, just a few nits remain, great work! :tada:  Make sure to add a detailed description when you open the pull request against the main repo! :rocket:

**Comment by guggero** [comment]
I think it makes sense to open the PR against the main repo now. I'll happily do a review there. Is that okay for everyone?

**Comment by bhandras** [comment]
> I think it makes sense to open the PR against the main repo now. I'll happily do a review there. Is that okay for everyone?

Yeah, sure. Actually we also just agree with @Vib-UX to open it against upstream, since the PR is quite ready as is.
